app/scripts/lib/pairing-channel-client.js

```diff
@@ -0,0 +1,170 @@
+   */
+  open (channelServerUri = this.get('channelServerUri'),
+    channelId = this.get('channelId'),
+    channelKey = this.get('channelKey'),
+  ) {
+    return this._importPairingChannel().then((FxAccountsPairingChannel) => {
+      if (this.channel) {
+        // to avoid opening a duplicate connection, say the client is connected
+        // if a socket exists but isn't yet connected.
+        throw PairingChannelClientErrors.toError('ALREADY_CONNECTED');
+      }
+
+      if (! channelServerUri || ! channelId || ! channelKey) {
+        throw PairingChannelClientErrors.toError('INVALID_CONFIGURATION');
+      }
+
+      const psk = base64urlToArrayBuffer(channelKey);
+      return FxAccountsPairingChannel.PairingChannel.connect(channelServerUri, channelId, psk).then((channel) => {
+
+        this.channel = channel;
+        this.set('isConnected', true);
+        this.trigger('connected');
+
+        this.channel.addEventListener('message', this._messageHandler.bind(this));
+        this.channel.addEventListener('error', this._errorHandler.bind(this));
+        this.channel.addEventListener('close', this._closeHandler.bind(this));
+
+      }).catch((err) => {
+        this.sentryMetrics.captureException(err);
+        this.trigger('error', err);
+      });
+    });
+  }
+
+  /**
+   * Close the channel.
+   *
+   * @returns {Promise} - rejects if no socket, resolves when connection closed
+   */
+  close () {
+    return Promise.resolve().then(() => {
+      if (! this.channel) {
+        throw PairingChannelClientErrors.toError('NOT_CONNECTED');
+      }
+
+      this.set('isConnected', false);
+      return this.channel.close();
+    });
+  }
+
+  /**
+   * Send `message` with `data` to the remote.
+   *
+   * @param {String} message name of message to send
+   * @param {Any} [data={}]
+   * @returns {Promise} resolves when complete
+   */
+  send (message, data = {}) {
+    if (! this.get('isConnected') || ! this.channel) {
+      return Promise.reject(PairingChannelClientErrors.toError('NOT_CONNECTED'));
+    }
+
+    if (! message) {
+      return Promise.reject(PairingChannelClientErrors.toError('INVALID_OUTBOUND_MESSAGE'));
+    }
+
+    return this.channel.send({
+      data,
+      message
+    });
+  }
+
+  /**
+   * Handle a message. Expects a message with the
+   * following format:
+   *
+   * ```
+   * event.detail = {
+   *   "data": {
+   *     "data": <message data>,
+   *     "message": <message name>,
+   *   },
+   *   "sender": {
+   *     "city": <sender's city name>,
+   *     "country": <sender's country name>,
+   *     "region": <sender's region>,
+   *     "ua": <sender's user agent string>,
+   *   }
+   * }
+   * ```
+   *
+   * An `error` message will be triggered if there is an error
+   * parsing or decrypting the message.
+   *
+   * @param {Object} event
+   * @private
+   */
+  _messageHandler (event) {
+    try {
+      const { data: payload, sender } = event.detail;
```

Comment:
> Since this data is coming in from outside of FxA, should we use a validation schema to ensure everything is of the shape we expect, e.g., that message is a string with length > 0, and that sender.city, country, region and UA are strings of some reasonable size? The reliers make use of schemas and it works well there. Vat doesn't support nested schemas, though calling validate multiple times is still possible.

https://github.com/mozilla/fxa-content-server/pull/6876/files#r253125816